lib/live_cache.rb

```diff
@@ -0,0 +1,145 @@
+# frozen_string_literal: true
+
+# This is a replacement for DistributedCache which is bounded and respects
+# clears during deploys
+class LiveCache
+  class Manager
+    CHANNEL_NAME ||= "/live_cache"
+
+    def initialize(message_bus: MessageBus)
+      @message_bus = message_bus
+      @subscription = Concurrent::Promises.delay(&method(:subscribe))
+      @mutex = Mutex.new
+      @caches = Hash.new { |h, k| h[k] = WeakList.new }
+    end
+
+    def register(hash_key, cache)
+      @mutex.synchronize { @caches[hash_key] << cache }
+
+      ensure_subscribed!
+    end
+
+    def publish(message)
+      @message_bus.publish(CHANNEL_NAME, message)
+    end
+
+    private
+
+    def process_message(message)
+      @mutex.synchronize do
+        @caches[message.data["hash_key"]].each { |cache| cache.process_message(message) }
+      end
+    end
+
+    def subscribe
+      @message_bus.subscribe(CHANNEL_NAME) { |message| process_message(message) }
+    end
+
+    def ensure_subscribed!
+      @subscription.wait!
+    end
+  end
+
+  class << self
+    attr_accessor :default_manager
+  end
+
+  self.default_manager = Manager.new
+
+  attr_reader :identity
+
+  def initialize(
+    hash_key,
+    max_size_per_site,
+    live_cache_multiplier: GlobalSetting.live_cache_multiplier,
+    manager: LiveCache.default_manager,
+    namespace: true
+  )
+    @manager = manager
+    @hash_key = hash_key
+    @data = SiteCache.new(live_cache_multiplier * max_size_per_site, max_size_per_site)
+    @manager.register(hash_key, self)
+    @identity = SecureRandom.hex
+    @namespace = namespace
+  end
+
+  def getset(key, &blk)
+    @data.getset(get_site_id, key, &blk)
+  end
+
+  def getset_bulk(ks, key_blk, &blk)
+    @data.getset_bulk(get_site_id, ks, key_blk, &blk)
+  end
+
+  def clear(after_commit: true)
```

Comment:
Because, in the vast majority of cases, we are writing to the cache because something changed in the DB. Doing it here once means we don't need to remember to do it in every case. This is the same as the `DistributedCache`:

https://github.com/discourse/discourse/blob/716e3a4dd58efde2d7d9f4691d9ca50b64991cc5/lib/distributed_cache.rb#L41-L47